[Writing Tools] Support compiling ENABLE_WRITING_TOOLS on the public SDK (Part 1)

rr-codes opened the pull request:
#### b504fc6a69adc63580f2c647dd2cff39a6a448fb
<pre>
[Writing Tools] Support compiling ENABLE_WRITING_TOOLS on the public SDK (Part 1)
<a href="https://bugs.webkit.org/show_bug.cgi?id=291790">https://bugs.webkit.org/show_bug.cgi?id=291790</a>
<a href="https://rdar.apple.com/149610431">rdar://149610431</a>

Reviewed by Abrar Rahman Protyasha.

Work towards making it possible to compile ENABLE_WRITING_TOOLS on the public SDK. This is the first
part of a series of commits.

In this PR, declare all the symbols of WritingTools and WritingToolsUI that WebKit uses when building
using the public SDK.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:

Add the symbols from WritingTools that WebKit uses. Also adds missing includes.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsUISPI.h:

Add the symbols from WritingToolsUI that WebKit uses. Also adds missing includes.

Canonical link: <a href="https://commits.webkit.org/293897@main">https://commits.webkit.org/293897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/887344fe973e4b15f12418d43c11c260feb92f13

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100223 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10170 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105353 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50806 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20177 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28346 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76301 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33360 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103230 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15429 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90524 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56661 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15249 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8523 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50174 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85156 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107712 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20031 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85254 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27699 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86731 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84793 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21547 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29472 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7207 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21205 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32506 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27084 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30400 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28643 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
```diff
@@ -34,25 +37,170 @@ DECLARE_SYSTEM_HEADER
 
-constexpr auto PlatformWritingToolsResultPlainText = NSWritingToolsResultPlainText;
-constexpr auto PlatformWritingToolsResultRichText = NSWritingToolsResultRichText;
-constexpr auto PlatformWritingToolsResultList = NSWritingToolsResultList;
-constexpr auto PlatformWritingToolsResultTable = NSWritingToolsResultTable;
+    WTRequestedToolProofread = 1,
+    WTRequestedToolRewrite = 2,
+    WTRequestedToolRewriteProofread = 3,
+
+    WTRequestedToolRewriteFriendly = 11,
+    WTRequestedToolRewriteProfessional = 12,
+    WTRequestedToolRewriteConcise = 13,
+    WTRequestedToolRewriteOpenEnded = 19,
+
+    WTRequestedToolTransformSummary = 21,
+    WTRequestedToolTransformKeyPoints = 22,
+    WTRequestedToolTransformList = 23,
+    WTRequestedToolTransformTable = 24,
+
+    WTRequestedToolSmartReply = 101,
+
+    WTRequestedToolCompose = 201,
```
I think we're not referring to most of these symbols directly, right? Do you want to pare down the list? And a couple are only used in API tests, so maybe those can just be bespoke forward declared in `Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm`?

rr-codes replied:
in a subsequent PR, I will actually be using all these (since rn I'm doing things in a not type safe way); I just haven't gotten around to it yet

aprotyas commented on Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
```diff
@@ -34,25 +37,170 @@ DECLARE_SYSTEM_HEADER
+
+// MARK: WTSession
+
+typedef NS_ENUM(NSInteger, WTCompositionSessionType) {
+    WTCompositionSessionTypeNone,
+    WTCompositionSessionTypeMagic,
+    WTCompositionSessionTypeFriendly,
+    WTCompositionSessionTypeProfessional,
+    WTCompositionSessionTypeConcise,
+    WTCompositionSessionTypeOpenEnded,
+    WTCompositionSessionTypeSummary,
+    WTCompositionSessionTypeKeyPoints,
+    WTCompositionSessionTypeList,
+    WTCompositionSessionTypeTable,
+    WTCompositionSessionTypeCompose,
+    WTCompositionSessionTypeSmartReply,
+    WTCompositionSessionTypeProofread,
+};
```
Same with these cases, many are not referred to directly in our project

rr-codes replied:
for this, I realize we're not using many, but I was hesitant about changing them since that would affect their ordinal number, so I didn't want to risk accidentally changing one of their numeric values

aprotyas commented on Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
```diff
@@ -34,25 +37,170 @@ DECLARE_SYSTEM_HEADER
+    WTActionShowOriginal = 1,
+    WTActionShowRewritten,
+    WTActionCompositionRestart,
+    WTActionCompositionRefine,
```
Unused, so we can skip

rr-codes replied:
this will be used eventually also 😅

aprotyas commented on Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
```diff
@@ -34,25 +37,170 @@ DECLARE_SYSTEM_HEADER
+
+@optional
+
+@property (nonatomic) BOOL wantsWritingToolsInlineEditing;
```
Unused

aprotyas commented on Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
```diff
@@ -34,25 +37,170 @@ DECLARE_SYSTEM_HEADER
+
+@property (nonatomic) BOOL wantsWritingToolsInlineEditing;
+
+- (BOOL)supportsWritingToolsAction:(WTAction)action;
```
Unused

aprotyas reviewed (commented):
Looked at WritingToolsSPI.h so far. It looks like we can omit some enum cases and interface properties since we're not referring to them.

rr-codes committed:
[Writing Tools] Support compiling ENABLE_WRITING_TOOLS on the public SDK (Part 1)
https://bugs.webkit.org/show_bug.cgi?id=291790
rdar://149610431

Reviewed by Abrar Rahman Protyasha.

Work towards making it possible to compile ENABLE_WRITING_TOOLS on the public SDK. This is the first
part of a series of commits.

In this PR, declare all the symbols of WritingTools and WritingToolsUI that WebKit uses when building
using the public SDK.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:

Add the symbols from WritingTools that WebKit uses. Also adds missing includes.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsUISPI.h:

Add the symbols from WritingToolsUI that WebKit uses. Also adds missing includes.

Canonical link: https://commits.webkit.org/293897@main
